### PR TITLE
Update starcoder example to use the docker compiler

### PR DESCRIPTION
**Pull request by GeorgesLorre**

### CODE REVIEW

File: fondant/compiler.py [hunk 1]
```diff
@@ -119,6 +119,11 @@ def _generate_spec(self, pipeline: Pipeline, extra_volumes: list) -> dict:
             # add in and out manifest paths to command
             command.extend(["--output_manifest_path", f"{path}/manifest.json"])
 
+            # add component spec to command if does not exist yet
+            if not component_op.arguments.get("component_spec"):
+                component_spec = component_op.component_spec.specification
+                command.extend(["--component_spec", json.dumps(component_spec)])
+
```
Comment (RobbeSneyders) [inline fondant/compiler.py:126]: I would move this to the `ComponentOp` class as this is not compiler specific.

File: examples/pipelines/starcoder/pipeline.py [hunk 2]
```diff
@@ -46,7 +48,20 @@
...
+            "$HOME/.config/gcloud/application_default_credentials.json:/root/.config/gcloud/application_default_credentials.json:ro"
+        ]
+        compiler.compile(pipeline=pipeline, extra_volumes=extra_volumes)
+        logger.info("Run `docker-compose up` to run the pipeline.")
 
-# compile
-client.compile_and_run(pipeline=pipeline)
+    else:
+        client = Client(host=PipelineConfigs.HOST)
+        client.compile_and_run(pipeline=pipeline)
```
Comment (RobbeSneyders) [inline examples/pipelines/starcoder/pipeline.py:67]: I would prefer to include this into the `Pipeline` class or a separate command before updating the example. Ideally, we don't need any code changes in this file to make it work with a different runner.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @GeorgesLorre!

File: examples/pipelines/starcoder/pipeline.py [hunk 3]
```diff
@@ -46,7 +48,20 @@
...
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--local", action="store_true")
+    if parser.parse_args().local:
+        compiler = DockerCompiler()
+        extra_volumes = [
+            "$HOME/.config/gcloud/application_default_credentials.json:/root/.config/gcloud/application_default_credentials.json:ro"
+        ]
+        compiler.compile(pipeline=pipeline, extra_volumes=extra_volumes)
+        logger.info("Run `docker-compose up` to run the pipeline.")
```
Comment (PhilippeMoussalli) [inline examples/pipelines/starcoder/pipeline.py:63]:
```suggestion
        logger.info("Run `docker compose up` to run the pipeline.")
```

File: examples/pipelines/starcoder/pipeline.py (hunk 2 above)
Comment (GeorgesLorre) [reply examples/pipelines/starcoder/pipeline.py:67]: Agreed, maybe as a different PR as I don't want to bloat this one

File: examples/pipelines/starcoder/pipeline.py (hunk 3 above)
Comment (PhilippeMoussalli) [reply examples/pipelines/starcoder/pipeline.py:63]: oh it seems like it works but I had to install it. never mind you can keep it as is

Comment (RobbeSneyders) [reply examples/pipelines/starcoder/pipeline.py:63]: It doesn't for me. `docker compose` is the recommended way ([link](https://stackoverflow.com/a/66516826/4098821))

File: fondant/compiler.py [hunk 4]
```diff
@@ -117,7 +122,17 @@ def _generate_spec(self, pipeline: Pipeline, extra_volumes: list) -> dict:
...
+                ]
+            )
+
+            # add component spec to command if does not exist yet
+            if not component_op.arguments.get("component_spec"):
+                component_spec = component_op.component_spec.specification
+                command.extend(["--component_spec", json.dumps(component_spec)])
 
             # add arguments if any to command
             for key, value in component_op.arguments.items():
```
Comment (PhilippeMoussalli) [inline fondant/compiler.py:138]: could you replace 

```python
for key, value in component_op.arguments.items():
    command.extend([f"--{key}", f"{value}"])
```

with 

```python
for key, value in component_op.arguments.items():
  command.extend([f"--{key}", f"{value}"])
  if isinstance(value, (dict, list)):
      command.extend([f"--{key}", json.dumps(value)])
  else:
      command.extend([f"--{key}", f"{value}"])
```

to fix parsing dictionary, list arguments?

Comment (GeorgesLorre) [reply fondant/compiler.py:133]: done

Comment (PhilippeMoussalli) [reply fondant/compiler.py:133]: thanks

**Review by PhilippeMoussalli — APPROVED** [review]
Thanks Georges!

**Pull request merged**